protocol/common/message_segmentation.go

```diff
@@ -0,0 +1,361 @@
+
+	segmentMessage := &SegmentMessage{
+		SegmentMessage: &protobuf.SegmentMessage{},
+	}
+	err := proto.Unmarshal(message.TransportLayer.Payload, segmentMessage.SegmentMessage)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal SegmentMessage")
+	}
+
+	logger.Debug("handling message segment", zap.String("EntireMessageHash", types.HexBytes(segmentMessage.EntireMessageHash).String()),
+		zap.Uint32("Index", segmentMessage.Index), zap.Uint32("SegmentsCount", segmentMessage.SegmentsCount))
+
+	alreadyCompleted, err := s.persistence.IsMessageAlreadyCompleted(segmentMessage.EntireMessageHash)
+	if err != nil {
+		return err
+	}
+	if alreadyCompleted {
+		return ErrMessageSegmentsAlreadyCompleted
+	}
+
+	if segmentMessage.SegmentsCount < 2 {
+		return ErrMessageSegmentsInvalidCount
+	}
+
+	err = s.persistence.SaveMessageSegment(segmentMessage, message.TransportLayer.SigPubKey, time.Now().Unix())
+	if err != nil {
+		return err
+	}
+
+	segments, err := s.persistence.GetMessageSegments(segmentMessage.EntireMessageHash, message.TransportLayer.SigPubKey)
+	if err != nil {
+		return err
+	}
+
+	if len(segments) != int(segmentMessage.SegmentsCount) {
+		return ErrMessageSegmentsIncomplete
+	}
+
+	// Combine payload
+	var entirePayload bytes.Buffer
+	for _, segment := range segments {
+		_, err := entirePayload.Write(segment.Payload)
+		if err != nil {
+			return errors.Wrap(err, "failed to write segment payload")
+		}
+	}
+
+	// Sanity check
+	entirePayloadHash := crypto.Keccak256(entirePayload.Bytes())
+	if !bytes.Equal(entirePayloadHash, segmentMessage.EntireMessageHash) {
+		return ErrMessageSegmentsHashMismatch
+	}
+
+	err = s.persistence.CompleteMessageSegments(segmentMessage.EntireMessageHash, message.TransportLayer.SigPubKey, time.Now().Unix())
+	if err != nil {
+		return err
+	}
+
+	message.TransportLayer.Payload = entirePayload.Bytes()
+
+	return nil
+}
+
+// SegmentationLayerV2 is capable of reconstructing the message from both complete and partial sets of data segments.
+// It has capability to perform forward error correction.
+func (s *MessageSender) handleSegmentationLayerV2(message *v1protocol.StatusMessage) error {
+	logger := s.logger.With(zap.String("site", "handleSegmentationLayerV2")).With(zap.String("hash", types.HexBytes(message.TransportLayer.Hash).String()))
+
+	segmentMessage := &SegmentMessage{
+		SegmentMessage: &protobuf.SegmentMessage{},
+	}
+	err := proto.Unmarshal(message.TransportLayer.Payload, segmentMessage.SegmentMessage)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal SegmentMessage")
+	}
+
+	logger.Debug("handling message segment", zap.String("EntireMessageHash", types.HexBytes(segmentMessage.EntireMessageHash).String()),
+		zap.Uint32("Index", segmentMessage.Index), zap.Uint32("SegmentsCount", segmentMessage.SegmentsCount),
+		zap.Uint32("ParitySegmentIndex", segmentMessage.ParitySegmentIndex), zap.Uint32("ParitySegmentsCount", segmentMessage.ParitySegmentsCount))
```

Comment:
I'd much appreciate if each argument was on a separate line. It's much easier to read, and it's also easier to modify later (git diffs will be clearer).

But not a big deal, just some style 🧂 

```suggestion
	logger.Debug("handling message segment", 
	    zap.String("EntireMessageHash", types.HexBytes(segmentMessage.EntireMessageHash).String()),
		zap.Uint32("Index", segmentMessage.Index), 
		zap.Uint32("SegmentsCount", segmentMessage.SegmentsCount),
		zap.Uint32("ParitySegmentIndex", segmentMessage.ParitySegmentIndex),
		zap.Uint32("ParitySegmentsCount", segmentMessage.ParitySegmentsCount))
```